src/platform_specific/affinity.unix.cc

```diff
@@ -1,20 +1,226 @@
+	if(ret != 0) {
+		CELERITY_WARN("Error retrieving initial process affinity mask. Unable to check whether enough logical cores are available to this process.{}",
+		    cfg.enabled ? " Will disable thread pinning." : "");
+		g_state.config.enabled = false;
+		return true;
+	}
+
+	// pinned threads per process: application, scheduler, executor, 1 device submitter per device if enabled
+	uint32_t pinned_threads_per_process = 3;
+	if(g_state.config.use_backend_device_submission_threads) { pinned_threads_per_process += g_state.config.num_devices; }
+	// total number of threads to be pinned across processes (legacy mode) - we assume that each process has been assigned the same number of device
+	const uint32_t total_threads = pinned_threads_per_process * g_state.config.num_legacy_processes;
+
+	if(g_state.config.enabled) {
+		// select the core set to use
+		std::vector<uint32_t> selected_core_ids = {};
+		if(!cfg.hardcoded_core_ids.empty()) {
+			// attempt to use the provided hardcoded IDs if they match the number of threads to be pinned
+			if(static_cast<uint32_t>(cfg.hardcoded_core_ids.size()) != total_threads) {
+				CELERITY_WARN("Hardcoded core ID count ({}) does not match the number of threads to be pinned ({}), downgrading to auto-pinning.",
+				    cfg.hardcoded_core_ids.size(), total_threads);
+			} else {
+				selected_core_ids = cfg.hardcoded_core_ids;
+			}
+		}
+		if(selected_core_ids.empty()) {
+			// otherwise, sequential core assignments for now; it is most important that each of the threads is "close"
+			// to the ones next to it in this sequence, so that communication between them is fast
+			selected_core_ids = get_available_sequential_cores(g_state.available_cores, total_threads, cfg.standard_core_start_id);
+			if(selected_core_ids.empty()) {
+				CELERITY_WARN("Insufficient available cores for thread pinning (required {}, {} available), disabling pinning.", //
+				    total_threads, CPU_COUNT(&g_state.available_cores));
+			}
+		}
+		// build our pinning plan based on the selected core list
+		if(selected_core_ids.empty()) {
+			g_state.config.enabled = false;
+		} else {
+			uint32_t current_core_id = cfg.legacy_process_index * pinned_threads_per_process;
+			g_state.thread_pinning_plan.emplace(thread_type::application, selected_core_ids[current_core_id++]);
+			g_state.thread_pinning_plan.emplace(thread_type::scheduler, selected_core_ids[current_core_id++]);
+			g_state.thread_pinning_plan.emplace(thread_type::executor, selected_core_ids[current_core_id++]);
+			if(g_state.config.use_backend_device_submission_threads) {
+				for(uint32_t i = 0; i < g_state.config.num_devices; ++i) {
+					const auto device_tid =
+					    static_cast<thread_type>(thread_type::first_device_submitter + i); // NOLINT(clang-analyzer-optin.core.EnumCastOutOfRange)
+					g_state.thread_pinning_plan.emplace(device_tid, selected_core_ids[current_core_id++]);
+				}
+			}
+		}
+	} else {
```

Comment:
```suggestion
	} else {
	CELERITY_DEBUG("Thread pinning is disabled.");
```

Probably wouldn't hurt to have, in case someone sets `CELERITY_THREAD_PINNING=off` in their environment and forgets about it.